modules/sharedIdSystem.js

```diff
@@ -0,0 +1,176 @@
+/**
+ * This module adds Shared ID support to the User ID module
+ * The {@link module:modules/userId} module is required.
+ * @module modules/sharedIdSystem
+ * @requires module:modules/userId
+ */
+import { submodule } from '../src/hook.js';
+import * as utils from '../src/utils.js';
+// These values should NEVER change. If
+// they do, we're no longer making ulids!
+const ENCODING = '0123456789ABCDEFGHJKMNPQRSTVWXYZ'; // Crockford's Base32
+const ENCODING_LEN = ENCODING.length;
+const TIME_MAX = Math.pow(2, 48) - 1;
+const TIME_LEN = 10;
+const RANDOM_LEN = 16;
+const MODULE_NAME = 'sharedId';
+
+/** @type {Submodule} */
+export const sharedIdSubmodule = {
+  /**
+   * used to link submodule with config
+   * @type {string}
+   */
+  name: MODULE_NAME,
+
+  sharedIdFactory: undefined,
+
+  /**
+   * decode the stored id value for passing to bid requests
+   * @function
+   * @param {string} value
+   * @returns {{sharedid:string}} or undefined if value doesn't exists
+   */
+  decode(value) {
+    return (value && typeof value['sharedid'] === 'string') ? { 'sharedid': value['sharedid'] } : undefined;
+  },
+
+  /**
+   * performs action to obtain id and return a value.
+   * @function
+   * @param {SubmoduleParams} [configParams]
+   * @returns {sharedId}
+   */
+  getId: function(configParams) {
+    let sharedId = this.sharedIdGenerator();
+    return {
+      id: sharedId
+    }
+  },
+  /**
+   * the shared id generator factory.
+   * @returns {*}
+   */
+  sharedIdGenerator: function () {
```

Comment:
this and below methods do not seem to be related to the submodule. It seems safe to extract them out of the module and just reuse it here. Even if we want to leave factory for submodule use, other methods may be still regular functions